src/openrct2-ui/windows/Staff.cpp

```diff
@@ -119,1240 +123,1147 @@ static rct_widget* window_staff_page_widgets[] = {
+        }
     }
 
-    if (w->height < w->min_height)
+    void OnViewportRotate() override
     {
-        w->height = w->min_height;
-        w->Invalidate();
+        switch (page)
+        {
+            case WINDOW_STAFF_OVERVIEW:
+                OverviewViewportRotate();
+                break;
+        }
     }
 
-    if (w->height > w->max_height)
+    void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
     {
-        w->Invalidate();
-        w->height = w->max_height;
+        switch (page)
+        {
+            case WINDOW_STAFF_OVERVIEW:
+                OverviewTextInput(widgetIndex, text);
+                break;
+        }
     }
 
-    rct_viewport* viewport = w->viewport;
-    if (viewport != nullptr)
+private:
+#pragma region Common events
+    void CommonMouseUp(rct_widgetindex widgetIndex)
     {
-        int32_t new_width = w->width - 30;
-        int32_t new_height = w->height - 62;
-
-        // Update the viewport size
-        if (viewport->width != new_width || viewport->height != new_height)
+        switch (widgetIndex)
         {
-            viewport->width = new_width;
-            viewport->height = new_height;
-            viewport->view_width = viewport->zoom.ApplyTo(new_width);
-            viewport->view_height = viewport->zoom.ApplyTo(new_height);
+            case WIDX_CLOSE:
+                Close();
+                break;
+            case WIDX_TAB_1:
+            case WIDX_TAB_2:
+            case WIDX_TAB_3:
+                SetPage(widgetIndex - WIDX_TAB_1);
+                break;
         }
     }
 
-    WindowStaffViewportInit(w);
-}
-
-/**
- * Handle the dropdown of patrol button.
- *  rct2: 0x006BDF98
- */
-void WindowStaffOverviewMousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget)
-{
-    switch (widgetIndex)
+    void CommonPrepareDrawBefore()
     {
-        case WIDX_LOCATE:
-            WindowStaffShowLocateDropdown(w, widget);
-            break;
-        case WIDX_PATROL:
-        {
-            // Dropdown names
-            gDropdownItems[0].Format = STR_SET_PATROL_AREA;
-            gDropdownItems[1].Format = STR_CLEAR_PATROL_AREA;
+        ColourSchemeUpdateByClass(this, static_cast<rct_windowclass>(WC_STAFF));
 
-            auto dropdownPos = ScreenCoordsXY{ widget->left + w->windowPos.x, widget->top + w->windowPos.y };
-            int32_t extray = widget->height() + 1;
-            WindowDropdownShowText(dropdownPos, extray, w->colours[1], 0, 2);
-            gDropdownDefaultIndex = 0;
+        if (window_staff_page_widgets[page] != widgets)
+        {
+            widgets = window_staff_page_widgets[page];
+            InitScrollWidgets();
+        }
+        SetPressedTab();
 
-            const auto peep = GetStaff(w);
-            if (peep == nullptr)
+        // Enable all widgets
+        for (rct_widgetindex widgetIndex = WIDX_TAB_1; widgetIndex != WIDX_OVERVIEW_COUNT && widgetIndex != WIDX_OPTIONS_COUNT;
```

Comment:
Keep this in the dedicated function that disabled the widgets as it makes things clearer.

Reply:
This is super dangerous as well and will walk passed the end of the array. You need to pass the count of the widgets directly or even better just set all the widgets to not disabled with `disabled_widgets = 0;`

Reply:
I changed it on IntelOrca's [recommendation](https://github.com/OpenRCT2/OpenRCT2/pull/17400#discussion_r901027021) so that any changes to disabled_widgets in the future wouldn't affect it. Is there any better way or should I leave it as = 0 ? 

For now I've reverted it to `disabled_widgets = 0;`

Note: this code is now in `DisableWidgets()`

Reply:
I didn't see Ted's comment. Hmm. I really don't like the way the loop has been done if you make it get the widx count from an array for the page that would be better.

Reply:
As the count is different for the Overview and Options tabs, I figured I'd make it more general and check whether it reaches WIDGETS_END. Let me know if there's a better way

Reply:
Yeah that's fine. You could just use widget type last instead of widgetsend.type but it's ultimately the same.